# 데이터 전처리
## #01. 패키지 설치

In [18]:
from pandas import read_csv, DataFrame, merge, concat
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sb

## #02. 데이터 구축

### 1. 데이터 파일 열기

In [4]:
df2018 = read_csv('./data/서울시 우리마을가게 상권분석서비스(상권-추정매출)_2018.csv', encoding='euc-kr')
df2019 = read_csv('./data/서울시 우리마을가게 상권분석서비스(상권-추정매출)_2019.csv', encoding='euc-kr')
df2020 = read_csv('./data/서울시 우리마을가게 상권분석서비스(상권-추정매출)_2020.csv', encoding='euc-kr')
df2021 = read_csv('./data/서울시 우리마을가게 상권분석서비스(상권-추정매출)_2021.csv', encoding='euc-kr')
df상권영역 = read_csv('./data/서울시 우리마을가게 상권분석서비스(상권영역).csv', encoding='euc-kr')
df건축물 = read_csv('./data/서울특별시 건축물대장 법정동 코드정보.csv', encoding='euc-kr')

### 2. 상권영역 데이터 구축

#### 상권 영역 데이터 확인

서울시에만 해당하는 데이터

In [5]:
df상권영역.head()

기준_년월_코드 상권_구분_코드 상권_구분_코드_명    상권_코드  ...  와이좌표_값  시군구_코드    행정동_코드  형태정보
0    201810        R       전통시장  1001453  ...  442084   11620  11620585   NaN
1    201810        R       전통시장  1001454  ...  442413   11620  11620595   NaN
2    201810        R       전통시장  1001474  ...  444227   11680  11680650   NaN
3    201810        R       전통시장  1001475  ...  443310   11680  11680670   NaN
4    201810        R       전통시장  1001412  ...  449328   11500  11500590   NaN

[5 rows x 10 columns]

#### 법정동 코드정보

전국에 대한 데이터

In [6]:
df건축물.head()

시군구코드  법정동코드  행정동코드  시도명 시군구명     법정동명 행정동명     적용시작일     적용만료일
0  42750  31224      0  강원도  영월군  산솔면 화원리  NaN  20211102  99991231
1  42750  31226      0  강원도  영월군  산솔면 석항리  NaN  20211102  99991231
2  42750  31221      0  강원도  영월군  산솔면 녹전리  NaN  20211102  99991231
3  42750  31223    312  강원도  영월군  산솔면 이목리  산솔면  20211102  99991231
4  42750  31225    312  강원도  영월군  산솔면 연상리  산솔면  20211102  99991231

In [12]:
df건축물['시군구코드'].value_counts()

시군구코드
47780    970
47740    921
47800    875
47810    835
43710    818
        ... 
27145      1
28265      1
25000      1
45118      1
22235      1
Name: count, Length: 500, dtype: int64

> `47780` 이라는 코드를 갖는 지역이 `970`곳이 됨을 확인할 수 있다. 즉, 코드값의 중복이 있다.

#### 시군구명, 시군구코드를 추출하여 중복데이터 제거

In [13]:
df시군구 = df건축물.filter(['시군구명', '시군구코드'])
df시군구 = df시군구.drop_duplicates()
df시군구.dropna(inplace=True)
df시군구

시군구명  시군구코드
0           영월군  42750
14          유성구  30200
23          평택시  41220
30      용인시 수지구  41465
32      용인시 처인구  41461
...         ...    ...
82870       대천시  44170
83192       온양시  44190
86860  동구 안심출장소  22145
88397     동해출장소  42105
92806  북구 칠곡출장소  22235

[476 rows x 2 columns]

#### 법정동 코드 정보를 상권영역에 결합

In [14]:
df상권영역2 = merge(df상권영역, df건축물.filter(['시군구명','시군구코드']), left_on='시군구_코드', right_on='시군구코드', how='inner')
df상권영역2

기준_년월_코드 상권_구분_코드 상권_구분_코드_명    상권_코드  ...    행정동_코드  형태정보  시군구명  시군구코드
0         201810        R       전통시장  1001453  ...  11620585   NaN   관악구  11620
1         201810        R       전통시장  1001453  ...  11620585   NaN   관악구  11620
2         201810        R       전통시장  1001453  ...  11620585   NaN   관악구  11620
3         201810        R       전통시장  1001453  ...  11620585   NaN   관악구  11620
4         201810        R       전통시장  1001453  ...  11620585   NaN   관악구  11620
...          ...      ...        ...      ...  ...       ...   ...   ...    ...
145707    201810        A       골목상권  1000381  ...  11350630   NaN   노원구  11350
145708    201810        A       골목상권  1000381  ...  11350630   NaN   노원구  11350
145709    201810        A       골목상권  1000381  ...  11350630   NaN   노원구  11350
145710    201810        A       골목상권  1000381  ...  11350630   NaN   노원구  11350
145711    201810        A       골목상권  1000381  ...  11350630   NaN   노원구  11350

[145712 rows x 12 columns]

#### 구에 대한 정보 확인

 얼마나 많은 구가 있는가?


In [15]:
df상권영역2['시군구명'].value_counts()

시군구명
종로구     14553
강남구     13915
영등포구    12284
성북구     11868
중구      11825
마포구      7739
성동구      7728
관악구      7568
동대문구     7268
용산구      5858
강서구      5313
구로구      4897
서초구      4760
서대문구     3936
강동구      3480
동작구      3250
은평구      3111
송파구      2925
양천구      2646
도봉구      2508
중랑구      2304
광진구      2040
강북구      2028
노원구      1170
금천구       738
Name: count, dtype: int64

#### 파생변수 생성

행정구는 값의 종류가 너무 많아서 더미변수화 하기가 어려울 것으로 보고 도심과 비도심으로 구분하는 파생변수를 생성
> 출처 : 서울2030플랜

In [16]:
도심 = ['종로구','중구','영등포구','서초구','강남구']

df상권영역2['도심'] = np.isin(df상권영역2['시군구명'], 도심)

df상권영역2.head()

기준_년월_코드 상권_구분_코드 상권_구분_코드_명    상권_코드  ... 형태정보  시군구명  시군구코드     도심
0    201810        R       전통시장  1001453  ...  NaN   관악구  11620  False
1    201810        R       전통시장  1001453  ...  NaN   관악구  11620  False
2    201810        R       전통시장  1001453  ...  NaN   관악구  11620  False
3    201810        R       전통시장  1001453  ...  NaN   관악구  11620  False
4    201810        R       전통시장  1001453  ...  NaN   관악구  11620  False

[5 rows x 13 columns]

#### 상권영역 데이터 저장

In [17]:
df상권영역2.to_excel('상권영역.xlsx', index=False)

### 3. 코로나 이전 매출 데이터

#### 단순데이터 결합

In [19]:
코로나이전df = concat([df2018, df2019])
코로나이전df.reset_index(drop=True, inplace=True)
코로나이전df

기준_년_코드  기준_분기_코드 상권_구분_코드  ... 연령대_50_매출_건수  연령대_60_이상_매출_건수 점포수
0          2018         1        D  ...            0               95   0
1          2018         1        D  ...            0                0  10
2          2018         1        A  ...            7                8   4
3          2018         1        A  ...           23                0   1
4          2018         1        A  ...            0                5   2
...         ...       ...      ...  ...          ...              ...  ..
298239     2019         4        R  ...            0                0   2
298240     2019         4        A  ...         3474             1319   2
298241     2019         4        A  ...        12869             5596   2
298242     2019         4        D  ...            0                0   6
298243     2019         4        A  ...         5590             1857   2

[298244 rows x 80 columns]

#### 골목상권 데이터만 추출

In [20]:
코로나이전df2 = 코로나이전df.query("상권_구분_코드_명 == '골목상권'")
코로나이전df2.head()

기준_년_코드  기준_분기_코드 상권_구분_코드  ... 연령대_50_매출_건수  연령대_60_이상_매출_건수 점포수
2     2018         1        A  ...            7                8   4
3     2018         1        A  ...           23                0   1
4     2018         1        A  ...            0                5   2
5     2018         1        A  ...          458              654   1
6     2018         1        A  ...         1930             1807   1

[5 rows x 80 columns]

#### 코로나 이전과 이후의 조사 범위를 동일하게 설정

코로나19 발생시점인 2020년 1월을 기준으로 2018년 1,2,3,4분기, 2019년 1,2,3,4분기 데이터와 2020년 1,2,3,4분기, 2021년 1,2분기 데이터만 수집된 상태.

그러므로 코로나 이전 데이터에서 2018년 1,2분기를 분석범위에서 제외해야 코로나 전후에 대한 데이터분포가 균형을 이루게 된다.

In [24]:
제외할index = 코로나이전df2.query("(기준_분기_코드==2018 and (기준_분기_코드==1 or 기준_분기_코드==2))").index
코로나이전df3 = 코로나이전df2.drop(제외할index)
코로나이전df3.reset_index(drop=True, inplace=True)
코로나이전df3.to_excel('코로나이전_매출데이터.xlsx', index=False)
코로나이전df3

기준_년_코드  기준_분기_코드 상권_구분_코드  ... 연령대_50_매출_건수  연령대_60_이상_매출_건수 점포수
0          2018         1        A  ...            7                8   4
1          2018         1        A  ...           23                0   1
2          2018         1        A  ...            0                5   2
3          2018         1        A  ...          458              654   1
4          2018         1        A  ...         1930             1807   1
...         ...       ...      ...  ...          ...              ...  ..
192482     2019         4        A  ...          211              464   3
192483     2019         4        A  ...          218              144   7
192484     2019         4        A  ...         3474             1319   2
192485     2019         4        A  ...        12869             5596   2
192486     2019         4        A  ...         5590             1857   2

[192487 rows x 80 columns]

### 4. 코로나 이후 매출데이터

In [23]:
코로나이후df = concat([df2020, df2021])
코로나이후df.reset_index(drop=True, inplace=True)

코로나이후df2 = 코로나이후df.query("상권_구분_코드_명 == '골목상권'")
코로나이후df2.to_excel('코로나이후_매출데이터.xlsx', index=False)
코로나이후df2

기준_년_코드  기준_분기_코드 상권_구분_코드  ... 연령대_50_매출_건수  연령대_60_이상_매출_건수 점포수
2          2020         1        A  ...            0                0   2
4          2020         1        A  ...         2614             1690  19
5          2020         3        A  ...            0                0   1
8          2020         1        A  ...           96               69   1
9          2020         1        A  ...          314              162   3
...         ...       ...      ...  ...          ...              ...  ..
197054     2021         1        A  ...         2990             2609   4
197055     2021         1        A  ...          566              256   3
197056     2021         1        A  ...          279              112   2
197057     2021         1        A  ...          269               97   1
197058     2021         1        A  ...         1717             1068  16

[126320 rows x 80 columns]